lib/logflare/backends/ingest_event_queue.ex

```diff
@@ -0,0 +1,288 @@
+defmodule Logflare.Backends.IngestEventQueue do
+  @moduledoc false
+  use GenServer
+  alias Logflare.Source
+  alias Logflare.Backends.Backend
+  require Ex2ms
+
+  @ets_table_mapper :ingest_event_queue_mapping
+  @ets_table :source_ingest_events
+
+  ## Server
+  def start_link(_args) do
+    GenServer.start_link(__MODULE__, [], name: __MODULE__, hibernate_after: 1_000)
+  end
+
+  def init(_args) do
+    :ets.new(@ets_table_mapper, [
+      :public,
+      :named_table,
+      :set,
+      {:write_concurrency, false},
+      {:read_concurrency, false},
+      {:decentralized_counters, false}
+    ])
+
+    {:ok, %{}}
+  end
+
+  @doc """
+  Retrieves a private tid of a given source-backend combination.
+  """
+  def get_tid({%Source{id: sid}, nil}), do: get_tid({sid, nil})
+  def get_tid({%Source{id: sid}, %Backend{id: bid}}), do: get_tid({sid, bid})
+
+  def get_tid({sid, bid}) do
+    :ets.lookup_element(@ets_table_mapper, {sid, bid}, 2, nil)
+    # staleness check
+    |> then(fn
+      nil ->
+        nil
+
+      tid ->
+        if(:ets.info(tid) != :undefined, do: tid)
+    end)
+  end
+
+  @doc """
+  Creates or updates a private :ets table. The :ets table mapper is stored in #{@ets_table_mapper} .
+  """
+  def upsert_tid({%Source{id: sid}, nil}), do: upsert_tid({sid, nil})
+  def upsert_tid({%Source{id: sid}, %Backend{id: bid}}), do: upsert_tid({sid, bid})
+
+  def upsert_tid({sid, bid} = sid_bid) when is_integer(sid) do
+    case get_tid(sid_bid) do
+      nil ->
+        # create and insert
+        tid =
+          :ets.new(@ets_table, [
+            :public,
+            :set,
+            {:decentralized_counters, false},
+            {:write_concurrency, true},
+            {:read_concurrency, true}
+          ])
+
+        :ets.insert(@ets_table_mapper, {{sid, bid}, tid})
+        {:ok, tid}
+
+      tid ->
+        {:error, :already_exists, tid}
+    end
+  end
+
+  @doc """
+  Retrieves the table size of a given tid
+  """
+  def get_table_size({%Source{id: sid}, nil}), do: get_table_size({sid, nil})
+  def get_table_size({%Source{id: sid}, %Backend{id: bid}}), do: get_table_size({sid, bid})
+
+  def get_table_size({sid, bid}) do
+    with tid when tid != nil <- get_tid({sid, bid}),
+         num when is_integer(num) <- :ets.info(tid, :size) do
```

Comment:
does this check if the table exists? otherwise, the is_integer/1 guard is always true

Reply:
it returns `:undefined` if the `:ets` table does not exist 
https://www.erlang.org/doc/apps/stdlib/ets.html#info/1

Reply:
I thought so! 👍